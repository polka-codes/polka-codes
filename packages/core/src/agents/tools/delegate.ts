import { z } from 'zod'

import { type AgentInfo, agentTool } from '../types'

export const delegateTool = (agents: AgentInfo[]) => {
  const agentInputSchemas = agents.map((agent) =>
    z.object({
      agentName: z.literal(agent.name),
      task: z.string().describe('The task to be completed by the target agent'),
      ...(agent.contextSchema ? { context: agent.contextSchema.describe('The context information for the task') } : {}),
    }),
  )

  let inputSchema: z.ZodSchema
  if (agents.length === 0) {
    throw new Error('At least one agent is required to delegate a task')
  }
  if (agents.length === 1) {
    inputSchema = agentInputSchemas[0]
  } else {
    inputSchema = z.union(agentInputSchemas as any)
  }

  return agentTool({
    id: 'delegate',
    description: 'Delegate a task to another agent to execute and receive the result back',
    inputSchema: inputSchema.describe('The input for the task'),
    outputSchema: z.object({
      status: z.enum(['success', 'failure']).describe('The status of the task'),
      output: z.string().describe('The output of the task'),
    }),
    execute: async (input, polka) => {
      try {
        const result = await polka.startTask(input.agentName, input.task, input.context)
        if (typeof result === 'string') {
          return {
            status: 'success' as const,
            output: result,
          }
        }
        return {
          status: 'failure' as const,
          output: result.error,
        }
      } catch (error: any) {
        return {
          status: 'failure' as const,
          output: 'message' in error ? error.message : (error?.toString() ?? JSON.stringify(error)),
        }
      }
    },
  })
}
