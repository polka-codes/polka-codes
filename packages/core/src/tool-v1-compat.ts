// generated by polka.codes
import { z } from 'zod'
import type { FullToolInfo, FullToolInfoV2, ToolParameter } from './tool'

function zodSchemaToParameters(schema: z.ZodObject<any>): ToolParameter[] {
  const parameters: ToolParameter[] = []
  const { shape } = schema

  for (const name in shape) {
    const def: z.ZodTypeAny = shape[name]
    const isOptional = def.safeParse(undefined).success

    const description = def.description || ''

    const param: ToolParameter = {
      name,
      description,
      required: !isOptional,
    }

    const usageValue = def.meta()?.usageValue as string | undefined
    if (usageValue) {
      param.usageValue = usageValue
    }

    if (def instanceof z.ZodObject) {
      param.children = zodSchemaToParameters(def)
    } else if (def instanceof z.ZodArray) {
      param.allowMultiple = true
      const element = def.element
      if (element instanceof z.ZodObject) {
        param.children = zodSchemaToParameters(element)
      }
    }

    parameters.push(param)
  }

  return parameters
}

export function toToolInfoV1(tool: FullToolInfoV2): FullToolInfo {
  const { parameters: zodSchema, ...rest } = tool

  const v1Parameters = zodSchemaToParameters(zodSchema)

  return {
    ...rest,
    parameters: v1Parameters,
  }
}
