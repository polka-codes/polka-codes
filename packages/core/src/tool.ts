export type ToolParameter = {
  name: string
  description: string
  required: boolean
  usageValue?: string
  allowMultiple?: boolean
  children?: ToolParameter[]
}

export type ToolParameterValue = string | { [key: string]: ToolParameterValue } | ToolParameterValue[]

export type ToolExample = {
  description: string
  parameters: { name: string; value: ToolParameterValue }[]
}

export enum PermissionLevel {
  // This tool is completely safe
  None = 0,
  // This tool can read files
  Read = 1,
  // This tool can write files
  Write = 2,
  // This tool can perform arbitrary action
  Arbitrary = 3,
}

export type ToolInfo = {
  name: string
  description: string
  parameters: ToolParameter[]
  examples?: ToolExample[]
  permissionLevel: PermissionLevel
}

export type FullToolInfo = ToolInfo & {
  handler: ToolHandler<ToolInfo, any>
  isAvailable: (provider: any) => boolean
}

export enum ToolResponseType {
  Reply = 'Reply',
  Exit = 'Exit',
  Invalid = 'Invalid',
  Error = 'Error',
  Interrupted = 'Interrupted',
  HandOver = 'HandOver',
  Delegate = 'Delegate',
  Pause = 'Pause',
}

// Reply to the tool use
export type ToolResponseReply = {
  type: ToolResponseType.Reply
  message: string
}

// Should end the message thread
// e.g. task completed successfully
export type ToolResponseExit = {
  type: ToolResponseType.Exit
  message: string
  object?: any
}

// The tool arguments are invalid
export type ToolResponseInvalid = {
  type: ToolResponseType.Invalid
  message: string
}

// Some error occurred when executing the tool
// e.g. network request error, IO error
export type ToolResponseError = {
  type: ToolResponseType.Error
  message: string
  // If true, the tool can be retried
  // e.g. network request error are generally retryable
  // but IO errors are not
  canRetry?: boolean
}

// The tool execution was interrupted
// e.g. user cancelled the tool execution
// or some security policy was violated
export type ToolResponseInterrupted = {
  type: ToolResponseType.Interrupted
  message: string
}

// Hand over the task to another agent
// e.g. hand over a coding task to the coder agent
export type ToolResponseHandOver = {
  type: ToolResponseType.HandOver
  agentName: string
  task: string
  context?: string
  files?: string[]
  originalTask?: string
}

// Delegate the task to another agent
export type ToolResponseDelegate = {
  type: ToolResponseType.Delegate
  agentName: string
  task: string
  context?: string
  files?: string[]
  originalTask?: string
}

export type ToolResponsePause = {
  type: ToolResponseType.Pause
  object: any
}

export type ToolResponse =
  | ToolResponseReply
  | ToolResponseExit
  | ToolResponseInvalid
  | ToolResponseError
  | ToolResponseInterrupted
  | ToolResponseHandOver
  | ToolResponseDelegate
  | ToolResponsePause

export type ToolHandler<T extends ToolInfo, P> = (
  provider: P,
  args: Partial<Record<T['parameters'][number]['name'], ToolParameterValue>>,
) => Promise<ToolResponse>
