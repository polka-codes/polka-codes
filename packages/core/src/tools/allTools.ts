export { default as askFollowupQuestion } from './askFollowupQuestion'
export { default as attemptCompletion } from './attemptCompletion'
export { default as executeCommand } from './executeCommand'
export { default as listCodeDefinitionNames } from './listCodeDefinitionNames'
export { default as listFiles } from './listFiles'
export { default as readFile } from './readFile'
export { default as replaceInFile } from './replaceInFile'
export { default as searchFiles } from './searchFiles'
export { default as writeToFile } from './writeToFile'
export { default as handOver } from './handOver'
export { default as removeFile } from './removeFile'
export { default as renameFile } from './renameFile'
