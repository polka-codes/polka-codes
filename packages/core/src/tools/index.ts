export * from './provider'
export * from './allTools'
export { default as replaceInFile } from './replaceInFile' // not included in allTools

export * as allTools from './allTools'
