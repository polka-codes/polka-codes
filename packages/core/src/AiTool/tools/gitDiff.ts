// generated by polka.codes
import { z } from 'zod'
import { type FullToolInfoV2, PermissionLevel, type ToolHandler, type ToolInfoV2, ToolResponseType } from '../../tool'
import type { CommandProvider } from '../../tools'

export const toolInfo = {
  name: 'git_diff',
  description:
    'Get the git diff for the current repository. Can be used to get staged changes, unstaged changes, or changes between commits. By default, it returns unstaged changes.',
  parameters: z.object({
    staged: z
      .preprocess((val) => {
        if (typeof val === 'string') {
          const lower = val.toLowerCase()
          if (lower === 'false') return false
          if (lower === 'true') return true
        }
        return val
      }, z.boolean().optional().default(false))
      .describe('Get staged changes instead of unstaged changes.'),
    commitRange: z.string().optional().describe('The commit range to get the diff for (e.g., "main...HEAD").'),
    file: z.string().optional().describe('Get the diff for a specific file.'),
  }),
  permissionLevel: PermissionLevel.Read,
} as const satisfies ToolInfoV2

export const handler: ToolHandler<typeof toolInfo, CommandProvider> = async (provider, args) => {
  if (!provider.executeCommand) {
    return {
      type: ToolResponseType.Error,
      message: 'Not possible to execute command. Abort.',
    }
  }

  const { staged, file, commitRange } = toolInfo.parameters.parse(args)

  const commandParts = ['git', 'diff', '--no-color']
  if (staged) {
    commandParts.push('--staged')
  }
  if (commitRange) {
    commandParts.push(commitRange)
  }
  if (file) {
    commandParts.push('--', file)
  }

  const command = commandParts.join(' ')
  try {
    const result = await provider.executeCommand(command, false)
    if (result.exitCode === 0) {
      if (!result.stdout.trim()) {
        return {
          type: ToolResponseType.Reply,
          message: 'No diff found.',
        }
      }
      return {
        type: ToolResponseType.Reply,
        message: `<diff file="${file ?? 'all'}">\n${result.stdout}\n</diff>`,
      }
    }
    return {
      type: ToolResponseType.Error,
      message: `\`${command}\` exited with code ${result.exitCode}:\n${result.stderr}`,
    }
  } catch (error) {
    return {
      type: ToolResponseType.Error,
      message: error instanceof Error ? error.message : String(error),
    }
  }
}

export const isAvailable = (provider: CommandProvider): boolean => {
  return !!provider.executeCommand
}

export default {
  ...toolInfo,
  handler,
  isAvailable,
} satisfies FullToolInfoV2
