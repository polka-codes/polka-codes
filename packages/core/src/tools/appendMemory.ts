import { z } from 'zod'
import { type FullToolInfo, type ToolHandler, type ToolInfo, ToolResponseType } from '../tool'
import type { MemoryProvider } from './provider'

export const toolInfo = {
  name: 'appendMemory',
  description: 'Appends content to a memory topic.',
  parameters: z.object({
    topic: z.string().optional().describe('The topic to append to in memory. Defaults to ":default:".'),
    content: z.string().describe('The content to append.'),
  }),
} as const satisfies ToolInfo

export const handler: ToolHandler<typeof toolInfo, MemoryProvider> = async (provider, args) => {
  const { topic, content } = toolInfo.parameters.parse(args)
  await provider.append(topic || ':default:', content)
  return {
    type: ToolResponseType.Reply,
    message: {
      type: 'text',
      value: `Content appended to memory topic '${topic || ':default:'}'.`,
    },
  }
}

export default {
  ...toolInfo,
  handler,
} satisfies FullToolInfo
