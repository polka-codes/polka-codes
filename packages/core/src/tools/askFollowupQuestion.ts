import { z } from 'zod'
import { type FullToolInfoV2, PermissionLevel, type ToolHandler, type ToolInfoV2, ToolResponseType } from '../tool'
import type { InteractionProvider } from './provider'

const questionObject = z.object({
  prompt: z.string().describe('The text of the question.').meta({ usageValue: 'question text here' }),
  options: z
    .array(z.string())
    .default([])
    .describe('Ordered list of suggested answers (omit if none).')
    .meta({ usageValue: 'suggested answer here' }),
})

export const toolInfo = {
  name: 'ask_followup_question',
  description:
    'Call this when vital details are missing. Pose each follow-up as one direct, unambiguous question. If it speeds the reply, add up to five short, mutually-exclusive answer options. Group any related questions in the same call to avoid a back-and-forth chain.',
  parameters: z
    .object({
      questions: z
        .array(questionObject)
        .describe('One or more follow-up questions you need answered before you can continue.')
        .meta({ usageValue: 'questions here' }),
    })
    .meta({
      examples: [
        {
          description: 'Single clarifying question (no options)',
          input: {
            questions: { prompt: 'What is the target deployment environment?' },
          },
        },
        {
          description: 'Single question with multiple-choice options',
          input: {
            questions: {
              prompt: 'Which frontend framework are you using?',
              options: ['React', 'Angular', 'Vue', 'Svelte'],
            },
          },
        },
        {
          description: 'Two related questions in one call',
          input: {
            questions: [
              { prompt: 'What type of application are you building?' },
              {
                prompt: 'Preferred programming language?',
                options: ['JavaScript', 'TypeScript', 'Python', 'Java'],
              },
            ],
          },
        },
        {
          description: 'Binary (yes/no) confirmation',
          input: {
            questions: {
              prompt: 'Is it acceptable to refactor existing tests to improve performance?',
              options: ['Yes', 'No'],
            },
          },
        },
      ],
    }),
  permissionLevel: PermissionLevel.None,
} as const satisfies ToolInfoV2

export const handler: ToolHandler<typeof toolInfo, InteractionProvider> = async (provider, args) => {
  if (!provider.askFollowupQuestion) {
    return {
      type: ToolResponseType.Error,
      message: 'Not possible to ask followup question. Abort.',
    }
  }

  const { questions } = toolInfo.parameters.parse(args)
  if (questions.length === 0) {
    return {
      type: ToolResponseType.Invalid,
      message: 'No questions provided',
    }
  }

  const answers = []
  for (const question of questions) {
    const { prompt, options } = question
    const answer = await provider.askFollowupQuestion(prompt, options)
    answers.push(`<ask_followup_question_answer question="${prompt}">
${answer}
</ask_followup_question_answer>`)
  }

  return {
    type: ToolResponseType.Reply,
    message: answers.join('\n'),
  }
}

export const isAvailable = (provider: InteractionProvider): boolean => {
  return !!provider.askFollowupQuestion
}

export default {
  ...toolInfo,
  handler,
  isAvailable,
} satisfies FullToolInfoV2
