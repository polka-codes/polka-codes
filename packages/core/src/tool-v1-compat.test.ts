// generated by polka.codes
import { describe, expect, it } from 'bun:test'
import { z } from 'zod'
import type { FullToolInfoV2 } from './tool'
import { ToolResponseType } from './tool'
import { toToolInfoV1 } from './tool-v1-compat'
import { askFollowupQuestion } from './tools'

describe('toToolInfoV1', () => {
  it('should convert a simple V2 tool to V1', () => {
    const toolV2: FullToolInfoV2 = {
      name: 'test-tool',
      description: 'A test tool',
      isAvailable: () => true,
      handler: async () => ({ type: ToolResponseType.Exit, message: 'test' }),
      parameters: z.object({
        param1: z.string().describe('Parameter 1'),
        param2: z.number().optional().describe('Parameter 2'),
      }),
    }

    const toolV1 = toToolInfoV1(toolV2)

    expect(toolV1.name).toBe('test-tool')
    expect(toolV1.description).toBe('A test tool')
    expect(toolV1.parameters).toMatchSnapshot()
  })

  it('should handle nested objects', () => {
    const toolV2: FullToolInfoV2 = {
      name: 'nested-tool',
      description: 'A tool with nested objects',
      isAvailable: () => true,
      handler: async () => ({ type: ToolResponseType.Exit, message: 'test' }),
      parameters: z.object({
        level1: z.object({
          level2: z.string().describe('Level 2 parameter'),
        }),
      }),
    }

    const toolV1 = toToolInfoV1(toolV2)

    expect(toolV1.parameters).toMatchSnapshot()
  })

  it('should handle arrays of strings', () => {
    const toolV2: FullToolInfoV2 = {
      name: 'array-tool',
      description: 'A tool with an array of strings',
      isAvailable: () => true,
      handler: async () => ({ type: ToolResponseType.Exit, message: 'test' }),
      parameters: z.object({
        items: z.array(z.string()).describe('List of items'),
      }),
    }

    const toolV1 = toToolInfoV1(toolV2)

    expect(toolV1.parameters).toMatchSnapshot()
  })

  it('should handle arrays of objects', () => {
    const toolV2: FullToolInfoV2 = {
      name: 'array-of-objects-tool',
      description: 'A tool with an array of objects',
      isAvailable: () => true,
      handler: async () => ({ type: ToolResponseType.Exit, message: 'test' }),
      parameters: z.object({
        objects: z
          .array(
            z.object({
              prop1: z.string().describe('Property 1'),
              prop2: z.boolean().describe('Property 2'),
            }),
          )
          .describe('List of objects'),
      }),
    }

    const toolV1 = toToolInfoV1(toolV2)

    expect(toolV1.parameters).toMatchSnapshot()
  })

  it('should match snapshot for a complex tool', () => {
    const toolV2: FullToolInfoV2 = {
      name: 'complex-tool',
      description: 'A complex tool for snapshot testing',
      isAvailable: () => true,
      handler: async () => ({ type: ToolResponseType.Exit, message: 'test' }),
      parameters: z.object({
        requiredString: z.string().describe('A required string'),
        optionalNumber: z.number().optional().describe('An optional number'),
        objectParam: z.object({
          nestedString: z.string().describe('A nested string'),
          nestedOptional: z.boolean().optional().describe('A nested optional boolean'),
        }),
        arrayString: z.array(z.string()).describe('An array of strings'),
        arrayObject: z
          .array(
            z.object({
              objProp: z.string().describe('An object property'),
            }),
          )
          .describe('An array of objects'),
      }),
    }

    const toolV1 = toToolInfoV1(toolV2)
    expect(toolV1).toMatchSnapshot()
  })

  it('should correctly convert a tool with nested optional arrays', () => {
    const toolV2: FullToolInfoV2 = askFollowupQuestion

    const toolV1 = toToolInfoV1(toolV2)

    expect(toolV1.parameters).toMatchSnapshot()
  })
})
