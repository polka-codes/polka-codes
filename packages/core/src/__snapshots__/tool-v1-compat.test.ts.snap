// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`toToolInfoV1 should convert a simple V2 tool to V1 1`] = `
[
  {
    "description": "Parameter 1",
    "name": "param1",
    "required": true,
  },
  {
    "description": "Parameter 2",
    "name": "param2",
    "required": false,
  },
]
`;

exports[`toToolInfoV1 should handle nested objects 1`] = `
[
  {
    "children": [
      {
        "description": "Level 2 parameter",
        "name": "level2",
        "required": true,
      },
    ],
    "description": "",
    "name": "level1",
    "required": true,
  },
]
`;

exports[`toToolInfoV1 should handle arrays of strings 1`] = `
[
  {
    "allowMultiple": true,
    "description": "List of items",
    "name": "items",
    "required": true,
  },
]
`;

exports[`toToolInfoV1 should handle arrays of objects 1`] = `
[
  {
    "allowMultiple": true,
    "children": [
      {
        "description": "Property 1",
        "name": "prop1",
        "required": true,
      },
      {
        "description": "Property 2",
        "name": "prop2",
        "required": true,
      },
    ],
    "description": "List of objects",
    "name": "objects",
    "required": true,
  },
]
`;

exports[`toToolInfoV1 should match snapshot for a complex tool 1`] = `
{
  "description": "A complex tool for snapshot testing",
  "handler": [Function: AsyncFunction],
  "isAvailable": [Function],
  "name": "complex-tool",
  "parameters": [
    {
      "description": "A required string",
      "name": "requiredString",
      "required": true,
    },
    {
      "description": "An optional number",
      "name": "optionalNumber",
      "required": false,
    },
    {
      "children": [
        {
          "description": "A nested string",
          "name": "nestedString",
          "required": true,
        },
        {
          "description": "A nested optional boolean",
          "name": "nestedOptional",
          "required": false,
        },
      ],
      "description": "",
      "name": "objectParam",
      "required": true,
    },
    {
      "allowMultiple": true,
      "description": "An array of strings",
      "name": "arrayString",
      "required": true,
    },
    {
      "allowMultiple": true,
      "children": [
        {
          "description": "An object property",
          "name": "objProp",
          "required": true,
        },
      ],
      "description": "An array of objects",
      "name": "arrayObject",
      "required": true,
    },
  ],
  "permissionLevel": 2,
}
`;

exports[`toToolInfoV1 should correctly convert a tool with nested optional arrays 1`] = `
[
  {
    "allowMultiple": true,
    "children": [
      {
        "description": "The text of the question.",
        "name": "prompt",
        "required": true,
        "usageValue": "question text here",
      },
      {
        "description": "Ordered list of suggested answers (omit if none).",
        "name": "options",
        "required": false,
        "usageValue": "suggested answer here",
      },
    ],
    "description": "One or more follow-up questions you need answered before you can continue.",
    "name": "questions",
    "required": true,
    "usageValue": "questions here",
  },
]
`;
